### ISSUE TITLE
Support for knowing which views are materialized

### ISSUE THREAD

**Issue by kgodey** [centerofci/mathesar#834]

**Comment by kgodey** [comment]
@dmos62 @ghislaineguerin @pavish @mathemancer @seancolsen @silentninja This issue has been updated and is ready for review. Please look through it and unassign yourself after you've added any feedback that you might have.

**Comment by silentninja** [comment]
So we don't support creating or modifying a materialized view but allow them to be read and used just like a normal view. Is that right?

**Comment by kgodey** [comment]
> So we don't support creating or modifying a materialized view but allow them to be read and used just like a normal view. Is that right?

Yes, that's correct.

**Comment by mathemancer** [comment]
Note that materialized views can't be `INSERT`ed into or `UPDATE`d under any circumstances (that I'm aware of; see https://www.postgresql.org/docs/13/rules-materializedviews.html). This means that some of our view logic won't work.  Given that, I think we should precisely specify what we mean by "used just like a normal view".  Unless we're ready to change their definitions and update them using the `REFRESH MATERIALIZED VIEW` command, I suggest we make them completely read-only for the moment, or at most allow changing the metadata, e.g., name, schema, owner, etc.

**Comment by pavish** [comment]
@ghislaineguerin I have a concern to keep in mind when we start work on the design for this.

I think materialized views should not be placed on the same level as normal views on the left pane, it might confuse the user since we do not support a lot of operations on materialized views for now. A clear distinction needs to be established before the user opens it, and not after.

**Comment by ghislaineguerin** [comment]
Looks good. I agree with @pavish in the need to differentiate the view type.

**Comment by kgodey** [comment]
@mathemancer Given that we won't be inserting or updating into views (see my comment here: https://github.com/centerofci/mathesar/issues/456#issuecomment-980290029), will we actually have less functionality for materialized views? It seems like the only difference will be that they might not have the latest data, whereas regular views will always have the latest data.

**Comment by mathemancer** [comment]
> @mathemancer Given that we won't be inserting or updating into views (see my comment here: [#456 (comment)](https://github.com/centerofci/mathesar/issues/456#issuecomment-980290029)), will we actually have less functionality for materialized views? It seems like the only difference will be that they might not have the latest data, whereas regular views will always have the latest data.

I think that's correct.  I didn't realize we were moving that direction for normal views (though I think that's a good move).  We might consider adding a way to refresh materialized views at some point, and the flow would be:

- click on cell
- record editor pops up; make desired changes to underlying table data
- click "refresh materialized view" button
- see changes in view.

One advantage of this over standard views would be that the user could defer updating the view until they'd done a number of edits.  In some cases, this could be a real time-saver.  I'm not sure the ability to refresh the view should be prioritized, but on the other hand, the implementation is literally one function that calls already-existing PostgreSQL functionality.  I'm not sure how we'd phrase it in the normal CRUD API functions, though.

**Comment by kgodey** [comment]
@mathemancer Thanks, makes sense. I've updated the description to include a refresh data action for materialized views.

@pavish @ghislaineguerin Given that we will not have reduced functionality for materialized views, I don't think we need to necessarily display them in a separate area. I'll put a note about them in #455 just in case.

**Comment by silentninja** [comment]
It would be really nice if the last refreshed timestamp is available for Materialized views. Since it isn't straightforward to get the information I wouldn't want it as a core feature but maybe as an extension or a plugin.

**Comment by kgodey** [comment]
@silentninja I agree but I don't think we should prioritize building that for the first release since our goal for the first release is just to display materialized views correctly rather than building in first-class support. I did add it to the "Design notes" section as something we might want to do in the future.
